common/SmokeTests/SmokeTest/Program.cs

```diff
@@ -1,37 +1,233 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
 using System;
-using System.Threading.Tasks;
-using SmokeTest.Samples;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
 
-namespace SmokeTest
+// The goal of the smoke tests is to ensure that each of the Azure SDK
+// packages can be used in the same project without causing conflicts.
+//
+// To validate, this project has been generated with references to all
+// packages for the current generation of the Azure SDK.  Each package assembly
+// and all of its transitive dependencies will be loaded, the types for each
+// enumerated, and instances will be created for any type with a parameterless constructor.
+//
+// This will ensure that all referenced packages have been loaded and can be used in
+// the same application without requiring specific knowledge of each SDK or code related to
+// its functionality.
+//
+// Execution will be logged to the console, using standard output for informational messages
+// and standard error for errors.   A return code of "0" will indicate a successful run; a
+// non-zero return indicates a test failure.
+
+const string AssemblyFileMask = "*.dll";
+
+LogInformation($"Smoke Test run starting...{ Environment.NewLine }");
+
+var returnCode = 0;
+
+// Starting with the entry assembly as the root, walk all references, both
+// direct and transient to process them.
+
+foreach (var assembly in LoadAssemblies(Assembly.GetEntryAssembly(), AssemblyFileMask))
 {
-    class Program
+    LogInformation($"Processing assembly: [{ assembly.FullName }]...");
+
+    var typeCount = 0;
+
+    try
     {
-        static async Task<int> Main(string[] args)
+        // Process each type in the assembly.  The call to "ProcessType" is safe and
+        // will not throw.
+
+        foreach (var type in assembly.GetTypes())
         {
-            Console.WriteLine("SMOKE TEST FOR TRACK 2 LIBRARIES");
-
-            await KeyVaultTest.RunTests();
-            await BlobStorageTest.RunTests();
-            await EventHubsTest.RunTests();
-            // Temporarily disable Cosmos DB smoke tests
-            // Issue to re-enable:
-            // https://github.com/Azure/azure-sdk-for-net/issues/11297
-            // await CosmosDBTest.RunTests();
-
-            Console.WriteLine();
-            Console.WriteLine("LIBRARY TESTS COMPLETE");
-            Console.WriteLine();
-            Console.WriteLine("SMOKE TEST FOR TRACK 2 SAMPLES");
-
-            await IotHubConnectionTests.RunTests();
-
-            Console.WriteLine("SAMPLES TESTS COMPLETE");
-            Console.WriteLine();
-            return 0;
+            ++typeCount;
+            ProcessType(type);
         }
     }
+    catch (Exception ex)
+    {
+        ReportError();
+        LogError($"{Environment.NewLine }[{ assembly.FullName }] could not be fully processed.  Error: { ex }{ Environment.NewLine }");
```

Comment:
Do we want to flag all the types with a problem in an assembly instead of stopping at the first failure?

Reply:
We don't stop on failure;  we write output about what type failed and keep going until we reach the end.     That said, we don't collect failures and do a nice "this is where the problems were"  - so that may be an improvement to the "go look through the output for errors" experience that we've got today.